### PR TITLE
Introduce response based cache expiry

### PR DESCRIPTION
**Pull request by kevinmade**

### CODE REVIEW

File: src/Contracts/Cacheable.php [hunk 1]
```diff
@@ -12,7 +15,7 @@ interface Cacheable
     public function resolveCacheDriver(): Driver;
 
     /**
-     * Define the cache expiry in seconds
+     * Resolve the cache expiry in seconds or as an DateTimeImmutable
      */
-    public function cacheExpiryInSeconds(): int;
+    public function resolveCacheExpiry(Response $response): DateTimeImmutable|int;
```
Comment (JonPurvis) [inline src/Contracts/Cacheable.php:20]: This will need to go out as a new major version (v4)

Comment (JonPurvis) [reply src/Contracts/Cacheable.php:20]: What happens if this returns 0 or < 0?

**Review by JonPurvis — CHANGES_REQUESTED** [review]
Great implementation! 

I had one question regarding `resolveCacheExpiry()` which I've commented in the file.

Other than that, could you add some tests that:

- Cover `DateTimeImmutable` being returned
- Assert the cached entry is written with the correct expiry
- Any edge cases in `resolveCacheExpiry` such as 0, negative, date in the past 

Thanks!

File: src/Contracts/Cacheable.php (hunk 1 above)
Comment (kevinmade) [reply src/Contracts/Cacheable.php:20]: If `resolveCacheExpiry()` returns 0 or a negative value, the response will not be cached at all. Only positive TTLs or future DateTimeImmutable values result in stored cache entries.

**Comment by JonPurvis** [comment]
Thanks for taking a look at my feedback, I'll get round to giving this a look over later this week / early next week. If it's all good, we can get it merged and get a new release out!

**Review by JonPurvis — APPROVED** [review]
👍 Looks good to me. 

I'll speak with @Sammyjo20 next week (if not before) to see if he's happy to get this merged and arrange a release of this package!

Thanks for your work on this!

**Comment by Sammyjo20** [comment]
The PR looks good to me but it still looks like it would be a breaking change because we're completely removing the `cacheExpiryInSeconds` method and changing it to `resolveCacheExpiry` which no one would have implemented, so updating the package would break it for their existing implementations.

**Comment by JonPurvis** [comment]
Hey @Sammyjo20 

We could perhaps release this as v4? We'd need a short upgrade guide somewhere though!

**Comment by kevinmade** [comment]
@Sammyjo20 Do you already have a rough idea for when v4 is planned to be released?

**Comment by JonPurvis** [comment]
@kevinmade Sam and myself are due a catch-up soon, so we'll discuss release plans then. There's a bunch of updates to release across the Saloon ecosystem, so we'll get them done around the same time.

**Comment by Sammyjo20** [comment]
Hey @kevinmade, thanks for your PR - I love this idea because it's common that websites will return a Cache header instructing how a response should be cached.

I had an idea I wanted validating with you to try and reduce the breaking changes. What if instead of removing the `cacheExpiryInSeconds` method, we implemented a caching mode? Something that can be set on the request or connector and allows us to specify if caching is driven manually (the way it works now) or automatically (your proposal).

Let me know your thoughts! 👍
